File: src/Client/Data/PaginatedList.ts
import {Pagination} from "./Pagination.js";

/**
 * A list with information relevant to the pagination of its items.
 */
export class PaginatedList<T> extends Array<T> {

	/**
	 * The information relevant to the pagination of the list items.
	 */
	pagination: Pagination;

	/**
	 * Creates a new paginated list.
	 * @param options An object providing values to initialize this instance.
	 */
	constructor(options: PaginatedListOptions<T> = {}) {
		super();
		for (const item of options.items ?? []) this.push(item);
		this.pagination = options.pagination ?? new Pagination;
	}

	/**
	 * Creates an empty paginated list.
	 * @param itemsPerPage The number of items per page.
	 * @returns An empty paginated list with the specified number of items per page.
	 */
	static empty<T>(itemsPerPage: number): PaginatedList<T> {
		return new this({pagination: new Pagination({itemsPerPage})});
	}
}

/**
 * Defines the options of a {@link PaginatedList} instance.
 */
export type PaginatedListOptions<T> = Partial<{

	/**
	 * The list items.
	 */
	items: T[];

	/**
	 * The information relevant to the pagination of the list items.
	 */
	pagination: Pagination;
}>;

File: src/Server/Base/Data/PaginatedList.cs
namespace Belin.Base.Data;

/// <summary>
/// A list with information relevant to the pagination of its items.
/// </summary>
/// <typeparam name="T">The type of items in the list.</typeparam>
public class PaginatedList<T>: List<T> {

	/// <summary>
	/// The information relevant to the pagination of the list items.
	/// </summary>
	public Pagination Pagination { get; set; } = new();

	/// <summary>
	/// Creates an empty paginated list.
	/// </summary>
	/// <param name="itemsPerPage">The number of items per page.</param>
	/// <returns>An empty paginated list with the specified number of items per page.</returns>
	public static PaginatedList<T> Empty(int itemsPerPage) => new() {
		Pagination = new() { ItemsPerPage = itemsPerPage }
	};
}
